components/raftstore-v2/src/operation/command/admin/merge/prepare.rs

```diff
@@ -276,6 +298,141 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+            req: Some(mem::take(req)),
+        });
+        Err(Error::PendingPrepareMerge)
+    }
+
+    pub fn merge_on_availability_response<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        from_peer: u64,
+        resp: &ExtraMessage,
+    ) {
+        let region_id = self.region_id();
+        let region_epoch = self.region().get_region_epoch().clone();
+        if self.merge_context().is_some()
+            && let Some(PrepareStatus::WaitForTrimStatus { pending_source, pending_target, req, .. }) = self
+                .merge_context_mut()
+                .prepare_status
+                .as_mut()
+            && req.is_some()
+        {
+            assert!(resp.has_availability_context());
+            if resp.get_availability_context().get_unavailable() {
+                self.take_merge_context();
+                return;
+            }
+            let from_region = resp.get_availability_context().get_from_region_id();
+            let from_epoch = resp.get_availability_context().get_from_region_epoch();
+            let trimmed = resp.get_availability_context().get_trimmed();
+            let target = req.as_ref().unwrap().get_admin_request().get_prepare_merge().get_target();
+            if  from_region == region_id && util::is_region_epoch_equal(from_epoch, &region_epoch)
+            {
+                if !trimmed {
+                    self.take_merge_context();
+                    return;
+                } else {
+                    pending_source.remove(&from_peer);
+                }
+            } else if from_region == target.get_id() && util::is_region_epoch_equal(from_epoch, target.get_region_epoch())
+            {
+                if !trimmed {
+                    self.take_merge_context();
+                    return;
+                } else {
+                    pending_target.remove(&from_peer);
+                }
+            }
+            if pending_source.is_empty() && pending_target.is_empty() {
+                let (ch, _) = CmdResChannel::pair();
+                let req = req.take().unwrap();
+                self.on_admin_command(store_ctx, req, ch);
+            }
+        }
+    }
+
+    fn already_checked_trim_status(&mut self) -> Result<Option<PreProposeContext>> {
+        match self
+            .merge_context()
+            .as_ref()
+            .and_then(|c| c.prepare_status.as_ref())
+        {
+            Some(PrepareStatus::WaitForTrimStatus {
+                pending_source,
+                pending_target,
+                ctx,
+                ..
+            }) => {
+                if pending_source.is_empty() && pending_target.is_empty() {
+                    Ok(Some(ctx.clone()))
+                } else {
+                    info!(
+                        self.logger,
+                        "reject PrepareMerge because trim status check is not complete";
+                    );
+                    Err(Error::PendingPrepareMerge)
```

Comment:
Maybe a metrics is needed for diagnostics?

Reply:
Log should be enough, this is not a frequent event.